library/general/disallowanonymous/template.yaml

```diff
@@ -0,0 +1,33 @@
+apiVersion: templates.gatekeeper.sh/v1beta1
+kind: ConstraintTemplate
+metadata:
+  name: k8sdisallowanonymous
+  annotations:
+    description: Disallows associating ClusterRole and Role resources to the system:anonymous user and system:unauthenticated group.
```

Comment:
Should we make this more generic by adding parameters that allow constraint authors to specify which users/groups should be disallowed?

Reply:
Thanks for reviewing @maxsmythe :)

I like the idea, I thought about way to implement the solution without over complicating the process for the constraint author. Came up with something like - 

```
apiVersion: constraints.gatekeeper.sh/v1beta1
kind: K8sDisallowAnonymous
metadata:
  name: no-anonymous
spec:
  match:
    kinds:
      - apiGroups: ["rbac.authorization.k8s.io"]
        kinds: ["ClusterRoleBinding"]
      - apiGroups: ["rbac.authorization.k8s.io"]
        kinds: ["RoleBinding"]
  parameters:
    roleMapping:
      - Users:
          - group-1
          - user-1 
        allowedRoles:
          - cluster-role-1
          - cluster-role-2
      - Users:
          - group-2
          - user-2
        allowedRoles:
          - cluster-role-3
          - cluster-role-4
```

I'm not sure if I'm confident with this though. This kind of ignores `ServiceAccounts` and the fact that they're namespace scoped. We might want the constraint to affect `ServiceAccounts` in specific namespaces (since their names may be the same across namespaces). On the other hand, taking the ServiceAccount Namespace into account will create a complex constraint that might be confusing for a constraint author.

Maybe the constraint should not be generic? After all, the original idea is to focus on an implementation of a specific "security guideline" that helps handling anonymous authentications (represented using the anonymous user and unauthenticated group).

Reply:
Yeah, there are a lot of edge cases depending on how generic we make this. Might be better to leave well enough alone for now.